### ISSUE TITLE
Uploading/adding product image to Azure blob fails

### ISSUE THREAD

**Issue by tombeuckelaere** [nopSolutions/nopCommerce#5050]
nopCommerce version: 4.3 
Clean install with the following options set:
- Hosted on Azure app service (also reproducable when running clean install localhost through visual studio)
- Blob storage settings are set in appsettings.json
- Media settings: file system

Steps to reproduce the problem: 

I'm getting the exception below when i'm trying to add an image to a product.
1. when uploading it, i can see the thumbnail in the storage container. The normal sized image isn't there
2. when i click "add picture" there is no thumbnail nor the normal sized image in the blob storage. Also the exception below is thrown when clicking the button.
3. tried both media settings "pictures are stored into database" and "pictures are stored into file system". Btw which one should it be when using azure blob storage since it isn't documented?

Thanks for your help and advice.

`Application: w3wp.exe CoreCLR Version: 4.700.20.41105 .NET Core Version: 3.1.8 Description: The process was terminated due to an unhandled exception. Exception Info: Microsoft.WindowsAzure.Storage.StorageException: The specified blob does not exist. at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.ExecuteAsyncInternal[T](RESTCommand``1 cmd, IRetryPolicy policy, OperationContext operationContext, CancellationToken token) at Nop.Services.Media.AzurePictureService.DeletePictureThumbsAsync(Picture picture) in D:\a\1\s\Libraries\Nop.Services\Media\AzurePictureService.cs:line 212 at Nop.Services.Media.AzurePictureService.DeletePictureThumbs(Picture picture) in D:\a\1\s\Libraries\Nop.Services\Media\AzurePictureService.cs:line 143 at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state) at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi) at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action``1 callback, TState& state) at System.Threading.QueueUserWorkItemCallback.Execute() at System.Threading.ThreadPoolWorkQueue.Dispatch() Request Information RequestID:a77839a3-501e-0039-0180-929c44000000 RequestDate:Thu, 24 Sep 2020 14:38:22 GMT StatusMessage:The specified blob does not exist. ErrorCode:BlobNotFound ErrorMessage:The specified blob does not exist. RequestId:a77839a3-501e-0039-0180-929c44000000 Time:2020-09-24T14:38:22.6174622Z`

**Comment by tombeuckelaere** [comment]
After applying https://github.com/nopSolutions/nopCommerce/pull/5019 with respect to the v4.3 code and also applying the fixes on that pull request, I've managed to get this functionality working.

For anyone who whishes to fix this on the current v4.3 source code, you'll find the correct working code below.
Don't forget to uninstall the nuget package `WindowsAzure.Storage` and to install the nuget package `Azure.Storage.Blobs`

`Nop.Services.Media.AzurePictureService`

```
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.IO;
using System.Threading.Tasks;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Microsoft.AspNetCore.Http;
using Nop.Core;
using Nop.Core.Caching;
using Nop.Core.Configuration;
using Nop.Core.Domain.Catalog;
using Nop.Core.Domain.Media;
using Nop.Core.Infrastructure;
using Nop.Data;
using Nop.Services.Caching;
using Nop.Services.Catalog;
using Nop.Services.Configuration;
using Nop.Services.Events;
using Nop.Services.Seo;

namespace Nop.Services.Media
{
    /// <summary>
    ///     Picture service for Windows Azure
    /// </summary>
    public class AzurePictureService : PictureService
    {
        private static bool _azureBlobStorageAppendContainerName;
        private static string _azureBlobStorageConnectionString;
        private static string _azureBlobStorageContainerName;
        private static string _azureBlobStorageEndPoint;
        private static BlobContainerClient _blobContainerClient;
        private static BlobServiceClient _blobServiceClient;
        private static bool _isInitialized;
        private readonly ICacheKeyService _cacheKeyService;
        private readonly object _locker = new object();
        private readonly MediaSettings _mediaSettings;
        private readonly IStaticCacheManager _staticCacheManager;

        public AzurePictureService(INopDataProvider dataProvider,
            IDownloadService downloadService,
            ICacheKeyService cacheKeyService,
            IEventPublisher eventPublisher,
            IHttpContextAccessor httpContextAccessor,
            INopFileProvider fileProvider,
            IProductAttributeParser productAttributeParser,
            IRepository<Picture> pictureRepository,
            IRepository<PictureBinary> pictureBinaryRepository,
            IRepository<ProductPicture> productPictureRepository,
            ISettingService settingService,
            IStaticCacheManager staticCacheManager,
            IUrlRecordService urlRecordService,
            IWebHelper webHelper,
            MediaSettings mediaSettings,
            NopConfig config)
            : base(dataProvider,
                downloadService,
                eventPublisher,
                httpContextAccessor,
                fileProvider,
                productAttributeParser,
                pictureRepository,
                pictureBinaryRepository,
                productPictureRepository,
                settingService,
                urlRecordService,
                webHelper,
                mediaSettings)
        {
            if (config is null)
                throw new ArgumentNullException(nameof(config));
            if (dataProvider is null)
                throw new ArgumentNullException(nameof(dataProvider));
            if (downloadService is null)
                throw new ArgumentNullException(nameof(downloadService));
            if (fileProvider is null)
                throw new ArgumentNullException(nameof(fileProvider));
            if (httpContextAccessor is null)
                throw new ArgumentNullException(nameof(httpContextAccessor));
            if (mediaSettings is null)
                throw new ArgumentNullException(nameof(mediaSettings));
            if (pictureBinaryRepository is null)
                throw new ArgumentNullException(nameof(pictureBinaryRepository));
            if (pictureRepository is null)
                throw new ArgumentNullException(nameof(pictureRepository));
            if (productAttributeParser is null)
                throw new ArgumentNullException(nameof(productAttributeParser));
            if (productPictureRepository is null)
                throw new ArgumentNullException(nameof(productPictureRepository));
            if (settingService is null)
                throw new ArgumentNullException(nameof(settingService));
            if (staticCacheManager is null)
                throw new ArgumentNullException(nameof(staticCacheManager));
            if (urlRecordService is null)
                throw new ArgumentNullException(nameof(urlRecordService));
            if (webHelper is null)
                throw new ArgumentNullException(nameof(webHelper));

            _cacheKeyService = cacheKeyService;
            _mediaSettings = mediaSettings;
            _staticCacheManager = staticCacheManager;

            OneTimeInit(config);
        }

        /// <summary>
        ///     Create cloud blob container
        /// </summary>
        protected virtual async Task CreateCloudBlobContainer()
        {
            await _blobContainerClient.CreateIfNotExistsAsync(PublicAccessType.Blob);
        }

        /// <summary>
        ///     Delete picture thumbs
        /// </summary>
        /// <param name="picture">Picture</param>
        protected override async void DeletePictureThumbs(Picture picture)
        {
            if (picture is null)
                throw new ArgumentNullException(nameof(picture));

            // create a string containing the blob name prefix
            var prefix = $"{picture.Id:0000000}";

            var tasks = new List<Task>();
            await foreach (var blob in _blobContainerClient.GetBlobsAsync(BlobTraits.All, BlobStates.All, prefix))
                tasks.Add(_blobContainerClient.DeleteBlobIfExistsAsync(blob.Name,
                    DeleteSnapshotsOption.IncludeSnapshots));

            await Task.WhenAll(tasks);

            _staticCacheManager.RemoveByPrefix(NopMediaDefaults.ThumbsExistsPrefixCacheKey);
        }

        /// <summary>
        ///     Get a value indicating whether some file (thumb) already exists
        /// </summary>
        /// <param name="thumbFilePath">Thumb file path</param>
        /// <param name="thumbFileName">Thumb file name</param>
        /// <returns>Result</returns>
        protected override bool GeneratedThumbExists(string thumbFilePath, string thumbFileName)
        {
            if (string.IsNullOrWhiteSpace(thumbFileName))
                return false;

            return GeneratedThumbExistsAsync(thumbFilePath, thumbFileName).GetAwaiter().GetResult();
        }

        /// <summary>
        ///     Initiates an asynchronous operation to get a value indicating whether some file (thumb)
        ///     already exists
        /// </summary>
        /// <param name="thumbFilePath">Thumb file path</param>
        /// <param name="thumbFileName">Thumb file name</param>
        /// <returns>Result</returns>
        [SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Ignored.")]
        protected virtual async Task<bool> GeneratedThumbExistsAsync(string thumbFilePath, string thumbFileName)
        {
            try
            {
                var key = _cacheKeyService.PrepareKeyForDefaultCache(NopMediaDefaults.ThumbExistsCacheKey,
                    thumbFileName);

                return await _staticCacheManager.GetAsync(key,
                    async () => { return await _blobContainerClient.GetBlobClient(thumbFileName).ExistsAsync(); });
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        ///     Get picture (thumb) local path
        /// </summary>
        /// <param name="thumbFileName">Filename</param>
        /// <returns>Local picture thumb path</returns>
        protected override string GetThumbLocalPath(string thumbFileName)
        {
            if (string.IsNullOrWhiteSpace(thumbFileName))
                return string.Empty;

            var path = _azureBlobStorageAppendContainerName ? $"{_azureBlobStorageContainerName}/" : string.Empty;

            return $"{_azureBlobStorageEndPoint}/{path}{thumbFileName}";
        }

        /// <summary>
        ///     Get picture (thumb) URL
        /// </summary>
        /// <param name="thumbFileName">Filename</param>
        /// <param name="storeLocation">
        ///     Store location URL; null to use determine the current store location automatically
        /// </param>
        /// <returns>Local picture thumb path</returns>
        protected override string GetThumbUrl(string thumbFileName, string storeLocation = null)
        {
            if (string.IsNullOrWhiteSpace(thumbFileName))
                return string.Empty;

            return GetThumbLocalPath(thumbFileName);
        }

        protected void OneTimeInit(NopConfig config)
        {
            if (_isInitialized)
                return;

            if (config is null)
                throw new ArgumentNullException(nameof(config));

            if (string.IsNullOrEmpty(config.AzureBlobStorageConnectionString))
                throw new Exception("Azure connection string for BLOB is not specified");

            if (string.IsNullOrEmpty(config.AzureBlobStorageContainerName))
                throw new Exception("Azure container name for BLOB is not specified");

            if (string.IsNullOrEmpty(config.AzureBlobStorageEndPoint))
                throw new Exception("Azure end point for BLOB is not specified");

            lock (_locker)
            {
                if (_isInitialized)
                    return;

                _azureBlobStorageAppendContainerName = config.AzureBlobStorageAppendContainerName;
                _azureBlobStorageConnectionString = config.AzureBlobStorageConnectionString;
                _azureBlobStorageContainerName = config.AzureBlobStorageContainerName.Trim().ToLower();
                _azureBlobStorageEndPoint = config.AzureBlobStorageEndPoint.Trim().ToLower().TrimEnd('/');

                _blobServiceClient = new BlobServiceClient(_azureBlobStorageConnectionString);
                _blobContainerClient = _blobServiceClient.GetBlobContainerClient(_azureBlobStorageContainerName);

                CreateCloudBlobContainer().GetAwaiter().GetResult(); // Run initialization synchronously.

                _isInitialized = true;
            }
        }

        /// <summary>
        ///     Save a value indicating whether some file (thumb) already exists
        /// </summary>
        /// <param name="thumbFilePath">Thumb file path</param>
        /// <param name="thumbFileName">Thumb file name</param>
        /// <param name="mimeType">MIME type</param>
        /// <param name="binary">Picture binary</param>
        protected override async void SaveThumb(string thumbFilePath, string thumbFileName, string mimeType,
            byte[] binary)
        {
            await SaveThumbAsync(thumbFilePath, thumbFileName, mimeType, binary);
        }

        /// <summary>
        ///     Initiates an asynchronous operation to save a value indicating whether some file (thumb)
        ///     already exists
        /// </summary>
        /// <param name="thumbFilePath">Thumb file path</param>
        /// <param name="thumbFileName">Thumb file name</param>
        /// <param name="mimeType">MIME type</param>
        /// <param name="binary">Picture binary</param>
        protected virtual async Task SaveThumbAsync(string thumbFilePath, string thumbFileName, string mimeType,
            byte[] binary)
        {
            var blobClient = _blobContainerClient.GetBlobClient(thumbFileName);
            using var ms = new MemoryStream(binary);

            BlobHttpHeaders headers = null;
            if (!string.IsNullOrWhiteSpace(mimeType))
                headers = new BlobHttpHeaders
                {
                    ContentType = mimeType
                };

            if (!string.IsNullOrWhiteSpace(_mediaSettings.AzureCacheControlHeader))
            {
                headers ??= new BlobHttpHeaders();
                headers.CacheControl = _mediaSettings.AzureCacheControlHeader;
            }

            if (headers is null)
                await blobClient.UploadAsync(ms);
            else
                await blobClient.UploadAsync(ms, new BlobUploadOptions {HttpHeaders = headers});

            _staticCacheManager.RemoveByPrefix(NopMediaDefaults.ThumbsExistsPrefixCacheKey);
        }
    }
}
```

**Comment by AndreiMaz** [comment]
@skoshelev Please also reply to https://www.nopcommerce.com/en/boards/topic/87849/uploadingadding-product-image-to-azure-blob-fails and check https://github.com/nopSolutions/nopCommerce/pull/5019

**Comment by skoshelev** [comment]
Fixed by #5019
